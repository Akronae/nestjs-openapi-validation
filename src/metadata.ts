/* eslint-disable */
export default async () => {
    const t = {
        ["./modules/app/app.dto"]: await import("./modules/app/app.dto")
    };
    return { "@nestjs/swagger": { "models": [[import("./modules/app/app.dto"), { "Query1": { str1: { required: true, type: () => String }, str2: { required: false, type: () => String }, date: { required: true, type: () => Date }, nbr1: { required: true, type: () => Number }, nbr2: { required: false, type: () => Number } }, "Query2": { str1: { required: true, type: () => String }, nbr1: { required: true, type: () => Number }, enum1: { required: true, type: () => Object } }, "Query3": { enum1: { required: true, type: () => Object }, enum2: { required: true, type: () => Object } }, "Query4": { query1: { required: true, type: () => t["./modules/app/app.dto"].Query1 }, field2: { required: true, type: () => Object } }, "Query5": { query4: { required: true, type: () => t["./modules/app/app.dto"].Query4 }, force: { required: true, type: () => Boolean }, query1: { required: false, type: () => t["./modules/app/app.dto"].Query1 } }, "Query6": { arr: { required: true, type: () => [t["./modules/app/app.dto"].Query4] } } }]], "controllers": [[import("./modules/app/app.controller"), { "AppController": { "getStatus": {}, "getStatusV2": { type: String }, "getQuery1": { type: t["./modules/app/app.dto"].Query1 }, "getQuery2": { type: t["./modules/app/app.dto"].Query2 }, "getQuery3": { type: t["./modules/app/app.dto"].Query3 }, "getQuery4": { type: t["./modules/app/app.dto"].Query4 }, "getQuery5": { type: t["./modules/app/app.dto"].Query5 }, "getQuery6": { type: t["./modules/app/app.dto"].Query6 } } }]] } };
};