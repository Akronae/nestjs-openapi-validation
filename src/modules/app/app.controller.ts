import { Body, Controller, Get, Post, Query, Version } from '@nestjs/common';
import {
  Query1,
  Query2,
  Query3,
  Query4,
  Query5,
  Query6,
  Query7,
  Query8,
  Query9,
} from './app.dto';
import { AppService } from './app.service';

@Controller({
  version: '1',
})
export class AppController {
  constructor(private readonly appService: AppService) {}

  @Get('status')
  getStatus() {
    return this.appService.getStatus();
  }

  @Get('status')
  @Version('2')
  getStatusV2() {
    return 'wow status version 2 wow';
  }

  @Get('query_1')
  getQuery1(@Query() query: Query1) {
    return query;
  }

  @Get('query_2')
  getQuery2(@Query() query: Query2) {
    return query;
  }

  @Get('query_3')
  getQuery3(@Query() query: Query3) {
    return query;
  }

  @Get('query_4')
  getQuery4(@Query() query: Query4) {
    return query;
  }

  @Post('query_5')
  getQuery5(@Body() query: Query5) {
    return query;
  }

  @Post('query_6')
  getQuery6(@Body() query: Query6) {
    return query;
  }

  @Post('query_7')
  getQuery7(@Body() query: Query7) {
    return query;
  }

  @Post('query_8')
  getQuery8(@Body() query: Query8) {
    return query;
  }

  @Post('query_9')
  getQuery9(@Body() query: Query9) {
    return query;
  }
}
