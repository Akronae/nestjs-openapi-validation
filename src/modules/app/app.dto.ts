import { ApiProperty, getSchemaPath } from '@nestjs/swagger';

export class Query1 {
  str1: string;
  str2?: string;
  date: Date;
  nbr1: number;
  nbr2?: number;
}

export enum Enum1Enum {
  A = 'A',
  B = 'B',
  C = 'C',
}
export type Enum1 = `${Enum1Enum}`;

export class Query2 {
  str1: string;
  nbr1: number;
  @ApiProperty({ enum: Enum1Enum })
  enum1: Enum1;
}

export enum Enum2Enum {
  AA = 'AA',
  BB = 'BB',
  CC = 'CC',
}
export type Enum2 = `${Enum2Enum}`;
export class Query3 {
  @ApiProperty({
    oneOf: [
      { enum: [...Object.values(Enum1Enum)] },
      { enum: [...Object.values(Enum2Enum)] },
    ],
  })
  enum1: Enum1 | Enum2;
  @ApiProperty({
    oneOf: [{ enum: [...Object.values(Enum1Enum)] }, { type: 'string' }],
  })
  enum2: Enum1 | (string & {});
}

export class Query4 {
  query1: Query1;
  @ApiProperty({
    oneOf: [{ $ref: getSchemaPath(Query2) }, { $ref: getSchemaPath(Query3) }],
  })
  field2: Query2 | Query3;
}

export class Query5 {
  query4: Query4;
  force: boolean;
  query1?: Query1;
}

export class Query6 {
  arr: Query4[];
}

export class Query7 {
  @ApiProperty({ minLength: 1, maxLength: 10 })
  str: string;
  @ApiProperty({ minimum: -3, maximum: 23 })
  nbr: number;
  @ApiProperty({ format: 'email' })
  email: string;
  @ApiProperty({ format: 'url' })
  url: string;
  @ApiProperty({ pattern: /^[0-9]{10}$/.source })
  phone: string;
}
