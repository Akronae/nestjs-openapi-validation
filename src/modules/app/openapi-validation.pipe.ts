import {
  ArgumentMetadata,
  BadRequestException,
  Injectable,
  PipeTransform,
} from '@nestjs/common';
import { OpenAPIObject } from '@nestjs/swagger';
import { mapEntries } from 'radash';
import { z, ZodType } from 'zod';

type OpenApiProp = {
  type?: string;
  enum?: [string, ...string[]];
  oneOf?: [OpenApiProp, OpenApiProp, ...OpenApiProp[]];
  $ref?: string;
};

@Injectable()
export class MetadataValidationPipe implements PipeTransform {
  private schemaMap: Record<
    string,
    Record<
      string,
      {
        type: () => { name: string };
        required: boolean;
      }
    >
  > = {};

  constructor(
    tsMetadata: () => Promise<{
      '@nestjs/swagger': {
        models: (
          | Promise<any>
          | Record<
              string,
              Record<
                string,
                { type: () => { name: string }; required: boolean }
              >
            >
        )[][];
      };
    }>,
    private readonly openapi: OpenAPIObject,
  ) {
    tsMetadata().then((res) => {
      const models = res['@nestjs/swagger'].models[0];
      for (const classes of Object.values(models)) {
        for (const [className, fields] of Object.entries(classes)) {
          this.schemaMap[className] = fields as (typeof this.schemaMap)[string];
        }
      }
    });
  }

  transform(value: any, metadata: ArgumentMetadata) {
    const dtoName = metadata.metatype?.name;
    if (!dtoName || !this.schemaMap[dtoName]) return value;

    const zodSchema = this.getZodSchema(dtoName);

    const res = zodSchema.safeParse(value);

    if (!res.success) {
      throw new BadRequestException(res.error);
    }

    return res.data;
  }

  getZodSchema(dtoName: string) {
    const tsschema = this.schemaMap[dtoName];
    const openapischema = this.openapi.components?.schemas[dtoName];
    // z.object({ a: z.coerce.().describe });

    const zodSchema = z.object({
      ...mapEntries(tsschema, (k, v) => {
        const prop = (openapischema as any).properties[k];
        const val = this.openapiPropToZod(prop, v);

        if (!val) {
          console.error(prop);
          throw new Error(`Unknown type: ${prop.type} for key ${k}`);
        }

        return [k, val];
      }),
    });

    return zodSchema;
  }

  openapiPropToZod(prop: OpenApiProp, opts: { required: boolean }): ZodType {
    if (prop.oneOf?.length > 1) {
      return z.union(
        prop.oneOf.map((o) => this.openapiPropToZod(o, { required: true })) as [
          ZodType,
          ZodType,
          ...ZodType[],
        ],
      );
    }

    if (prop.$ref) {
      const dtoname = prop.$ref.split('/').pop();
      return this.getZodSchema(dtoname);
    }

    const type = prop.type;

    let val = prop.enum
      ? z.enum(prop.enum)
      : type in z.coerce
        ? z.coerce[type]()
        : null;

    if (!val) {
      return null;
    }
    if (opts.required === false) val = val.optional();
    else val = requiredrefine(val, type);

    return val as ZodType;
  }
}

const requiredrefine = (x: any, type: string) =>
  x.superRefine((data, ctx) => {
    if (data == null || data == '' || data == 'undefined') {
      ctx.addIssue({
        code: z.ZodIssueCode.invalid_type,
        expected: type,
        received: data,
        message: 'required',
      });
    }
  });
