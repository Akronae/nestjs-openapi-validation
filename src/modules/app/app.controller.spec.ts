import {
  INestApplication,
  ValidationPipe,
  VersioningType,
} from '@nestjs/common';
import { DocumentBuilder, SwaggerModule } from '@nestjs/swagger';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';
import metadata from '../../../src/metadata';
import { OpenApiValidationPipe } from '../../lib/openapi-validation.pipe';
import { Query6, Query7, Query8, Query9 } from './app.dto';
import { AppModule } from './app.module';
import { AppService } from './app.service';

describe('AppController (e2e)', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const mod: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
      providers: [AppService],
    }).compile();
    app = mod.createNestApplication();

    app.enableVersioning({
      type: VersioningType.URI,
      defaultVersion: '1',
    });

    const config = new DocumentBuilder()
      .setTitle('API docs')
      .setDescription('The API description')
      .setVersion('1.0')
      .build();

    await SwaggerModule.loadPluginMetadata(metadata);
    const document = SwaggerModule.createDocument(app, config);
    SwaggerModule.setup('api', app, document);

    app.useGlobalPipes(
      new OpenApiValidationPipe(metadata, document),
      new ValidationPipe({
        transform: true,
        transformOptions: { enableImplicitConversion: true },
      }),
    );

    await app.init();
  });

  it('/query_1 fail (GET)', async () => {
    const res = await request(app.getHttpServer()).get('/v1/query_1');
    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_1 success (GET)', async () => {
    const res = await request(app.getHttpServer())
      .get('/v1/query_1')
      .query({
        str1: 'lala',
        date: new Date('2025'),
        nbr1: 123,
      });
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_2 fail (GET)', async () => {
    const res = await request(app.getHttpServer()).get('/v1/query_2').query({
      str1: 'lala',
    });
    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_2 success (GET)', async () => {
    const res = await request(app.getHttpServer()).get('/v1/query_2').query({
      str1: 'lala',
      nbr1: 123,
      enum1: 'A',
    });
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_3 fail (GET)', async () => {
    const res = await request(app.getHttpServer()).get('/v1/query_3').query({
      str1: 'lala',
    });
    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_3 success (GET)', async () => {
    const res = await request(app.getHttpServer()).get('/v1/query_3').query({
      str1: 'lala',
      enum2: 'okok',
      enum1: 'BB',
    });
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_4 fail (GET)', async () => {
    const res = await request(app.getHttpServer())
      .get('/v1/query_4')
      .query({
        query1: { a: 1 },
        query2: { a: 1 },
      });
    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_4 success (GET)', async () => {
    const res = await request(app.getHttpServer())
      .get('/v1/query_4')
      .query({
        query1: {
          str1: 'ahah',
          date: new Date('2025-05-02'),
          nbr1: 99,
        },
        field2: {
          str1: '!!!',
          enum2: 'okok',
          enum1: 'A',
        },
      });
    expect(res.status).toBe(200);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_5 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_5')
      .send({
        query4: {
          query1: {
            str1: 'ahah',
            date: new Date('2025-05-02'),
            nbr1: 99,
          },
          field2: {
            str1: '!!!',
            enum2: 'okok',
            enum1: 'A',
          },
        },
        force: false,
      });
    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_5 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_5')
      .send({
        query4: {
          query1: {
            str1: 'ahah',
            date: '2025-05-aa',
            nbr1: '99',
          },
          field2: {
            str1: '!!!',
            enum2: 'okok',
            enum1: 'D',
          },
        },
        force: true,
      });
    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_6 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_6')
      .send({
        arr: [
          {
            field2: { enum1: 'A', str1: 'lala', nbr1: 123 },
            query1: { str1: 'lalala', date: new Date('2025-05-02'), nbr1: 999 },
          },
          {
            field2: { enum1: 'AA', enum2: 'Aahah' },
            query1: { str1: 'lalala', date: new Date('2025-10-02'), nbr1: 111 },
          },
        ],
      } satisfies Query6);

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_6 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_6')
      .send({
        arr: [
          {
            field2: { enum1: 'A', str1: 'lala', nbr1: 123 },
            query1: { str1: 'lalala', date: new Date('2025-05-02'), nbr1: 999 },
          },
          {
            field2: { enum1: 'AAA', enum2: 'A' },
            query1: { str1: 'lalala', date: new Date('2025-10-02'), nbr1: 111 },
          },
        ],
      });
    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_7 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_7')
      .send({
        str: 'hello',
        nbr: 14,
        email: 'hey@hey.com',
        url: 'https://google.com',
        phone: '1234567894',
      } satisfies Query7);

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_7 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_7')
      .send({
        str: 'a',
        nbr: -10,
        email: 'hey',
        url: 'google',
        phone: '12',
      } satisfies Query7);

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_7 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_7')
      .send({
        str: 'hellooooooooooooooooooooooooooooooooooo',
        nbr: 140000,
        email: 'hey@hey.com',
        url: 'https://google.com',
        phone: '1234567894',
      } satisfies Query7);

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_8 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_8')
      .send({
        nested: {
          long: {
            prop: 21,
          },
        },
      } satisfies Query8);

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_8 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_8')
      .send({
        nested: {
          long: {
            prop: 'BB' as any,
          },
        },
      } satisfies Query8);

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });

  it('/query_9 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
          semi: {},
        },
      } satisfies Query9);

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
        },
      } satisfies Query9);

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer()).post('/v1/query_9').send({});

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer()).post('/v1/query_9').send({
      required: {},
    });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: 31,
        },
      });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 'dd' },
        },
      });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
        },
        opt: {},
      });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
        },
        opt: { long: {} },
      });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
        },
        opt: { long: { prop: 'dsdd' } },
      });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 fail (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
          opt: {},
        },
      });

    expect(res.status).toBe(400);
    expect(res.body).toMatchSnapshot();
  });
  it('/query_9 success (POST)', async () => {
    const res = await request(app.getHttpServer())
      .post('/v1/query_9')
      .send({
        required: {
          long: { prop: 21 },
        },
        opt: { long: { prop: 21 } },
      } as Query9);

    expect(res.status).toBe(201);
    expect(res.body).toMatchSnapshot();
  });
});
