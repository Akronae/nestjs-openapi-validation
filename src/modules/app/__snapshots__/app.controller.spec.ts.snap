// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`AppController (e2e) /query_1 fail (GET) 1`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "required",
      "path": [
        "str1",
      ],
      "received": "undefined",
    },
    {
      "code": "invalid_string",
      "message": "Invalid datetime",
      "path": [
        "date",
      ],
      "validation": "datetime",
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "required",
      "path": [
        "date",
      ],
      "received": "undefined",
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "nbr1",
      ],
      "received": "nan",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_1 success (GET) 1`] = `
{
  "date": "2025-01-01T00:00:00.000Z",
  "nbr1": 123,
  "str1": "lala",
}
`;

exports[`AppController (e2e) /query_2 fail (GET) 1`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "nbr1",
      ],
      "received": "nan",
    },
    {
      "code": "invalid_type",
      "expected": "'A' | 'B' | 'C'",
      "message": "Required",
      "path": [
        "enum1",
      ],
      "received": "undefined",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_2 success (GET) 1`] = `
{
  "enum1": "A",
  "nbr1": 123,
  "str1": "lala",
}
`;

exports[`AppController (e2e) /query_3 fail (GET) 1`] = `
{
  "issues": [
    {
      "code": "invalid_union",
      "message": "Invalid input",
      "path": [
        "enum1",
      ],
      "unionErrors": [
        {
          "issues": [
            {
              "code": "invalid_type",
              "expected": "'A' | 'B' | 'C'",
              "message": "Required",
              "path": [
                "enum1",
              ],
              "received": "undefined",
            },
          ],
          "name": "ZodError",
        },
        {
          "issues": [
            {
              "code": "invalid_type",
              "expected": "'AA' | 'BB' | 'CC'",
              "message": "Required",
              "path": [
                "enum1",
              ],
              "received": "undefined",
            },
          ],
          "name": "ZodError",
        },
      ],
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "required",
      "path": [
        "enum2",
      ],
      "received": "undefined",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_3 success (GET) 1`] = `
{
  "enum1": "BB",
  "enum2": "okok",
}
`;

exports[`AppController (e2e) /query_4 fail (GET) 1`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "required",
      "path": [
        "query1",
        "str1",
      ],
      "received": "undefined",
    },
    {
      "code": "invalid_string",
      "message": "Invalid datetime",
      "path": [
        "query1",
        "date",
      ],
      "validation": "datetime",
    },
    {
      "code": "invalid_type",
      "expected": "string",
      "message": "required",
      "path": [
        "query1",
        "date",
      ],
      "received": "undefined",
    },
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "query1",
        "nbr1",
      ],
      "received": "nan",
    },
    {
      "code": "invalid_union",
      "message": "Invalid input",
      "path": [
        "field2",
      ],
      "unionErrors": [
        {
          "issues": [
            {
              "code": "invalid_type",
              "expected": "object",
              "message": "Required",
              "path": [
                "field2",
              ],
              "received": "undefined",
            },
          ],
          "name": "ZodError",
        },
        {
          "issues": [
            {
              "code": "invalid_type",
              "expected": "object",
              "message": "Required",
              "path": [
                "field2",
              ],
              "received": "undefined",
            },
          ],
          "name": "ZodError",
        },
      ],
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_4 success (GET) 1`] = `
{
  "field2": {
    "enum1": "A",
    "enum2": "okok",
  },
  "query1": {
    "date": "2025-05-02T00:00:00.000Z",
    "nbr1": 99,
    "str1": "ahah",
  },
}
`;

exports[`AppController (e2e) /query_5 fail (POST) 1`] = `
{
  "issues": [
    {
      "code": "invalid_string",
      "message": "Invalid datetime",
      "path": [
        "query4",
        "query1",
        "date",
      ],
      "validation": "datetime",
    },
    {
      "code": "invalid_union",
      "message": "Invalid input",
      "path": [
        "query4",
        "field2",
      ],
      "unionErrors": [
        {
          "issues": [
            {
              "code": "invalid_type",
              "expected": "number",
              "message": "Expected number, received nan",
              "path": [
                "query4",
                "field2",
                "nbr1",
              ],
              "received": "nan",
            },
            {
              "code": "invalid_enum_value",
              "message": "Invalid enum value. Expected 'A' | 'B' | 'C', received 'D'",
              "options": [
                "A",
                "B",
                "C",
              ],
              "path": [
                "query4",
                "field2",
                "enum1",
              ],
              "received": "D",
            },
          ],
          "name": "ZodError",
        },
        {
          "issues": [
            {
              "code": "invalid_union",
              "message": "Invalid input",
              "path": [
                "query4",
                "field2",
                "enum1",
              ],
              "unionErrors": [
                {
                  "issues": [
                    {
                      "code": "invalid_enum_value",
                      "message": "Invalid enum value. Expected 'A' | 'B' | 'C', received 'D'",
                      "options": [
                        "A",
                        "B",
                        "C",
                      ],
                      "path": [
                        "query4",
                        "field2",
                        "enum1",
                      ],
                      "received": "D",
                    },
                  ],
                  "name": "ZodError",
                },
                {
                  "issues": [
                    {
                      "code": "invalid_enum_value",
                      "message": "Invalid enum value. Expected 'AA' | 'BB' | 'CC', received 'D'",
                      "options": [
                        "AA",
                        "BB",
                        "CC",
                      ],
                      "path": [
                        "query4",
                        "field2",
                        "enum1",
                      ],
                      "received": "D",
                    },
                  ],
                  "name": "ZodError",
                },
              ],
            },
          ],
          "name": "ZodError",
        },
      ],
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_5 success (POST) 1`] = `
{
  "force": false,
  "query4": {
    "field2": {
      "enum1": "A",
      "enum2": "okok",
    },
    "query1": {
      "date": "2025-05-02T00:00:00.000Z",
      "nbr1": 99,
      "str1": "ahah",
    },
  },
}
`;

exports[`AppController (e2e) /query_6 fail (POST) 1`] = `
{
  "issues": [
    {
      "code": "invalid_union",
      "message": "Invalid input",
      "path": [
        "arr",
        1,
        "field2",
      ],
      "unionErrors": [
        {
          "issues": [
            {
              "code": "invalid_type",
              "expected": "string",
              "message": "required",
              "path": [
                "arr",
                1,
                "field2",
                "str1",
              ],
              "received": "undefined",
            },
            {
              "code": "invalid_type",
              "expected": "number",
              "message": "Expected number, received nan",
              "path": [
                "arr",
                1,
                "field2",
                "nbr1",
              ],
              "received": "nan",
            },
            {
              "code": "invalid_enum_value",
              "message": "Invalid enum value. Expected 'A' | 'B' | 'C', received 'AAA'",
              "options": [
                "A",
                "B",
                "C",
              ],
              "path": [
                "arr",
                1,
                "field2",
                "enum1",
              ],
              "received": "AAA",
            },
          ],
          "name": "ZodError",
        },
        {
          "issues": [
            {
              "code": "invalid_union",
              "message": "Invalid input",
              "path": [
                "arr",
                1,
                "field2",
                "enum1",
              ],
              "unionErrors": [
                {
                  "issues": [
                    {
                      "code": "invalid_enum_value",
                      "message": "Invalid enum value. Expected 'A' | 'B' | 'C', received 'AAA'",
                      "options": [
                        "A",
                        "B",
                        "C",
                      ],
                      "path": [
                        "arr",
                        1,
                        "field2",
                        "enum1",
                      ],
                      "received": "AAA",
                    },
                  ],
                  "name": "ZodError",
                },
                {
                  "issues": [
                    {
                      "code": "invalid_enum_value",
                      "message": "Invalid enum value. Expected 'AA' | 'BB' | 'CC', received 'AAA'",
                      "options": [
                        "AA",
                        "BB",
                        "CC",
                      ],
                      "path": [
                        "arr",
                        1,
                        "field2",
                        "enum1",
                      ],
                      "received": "AAA",
                    },
                  ],
                  "name": "ZodError",
                },
              ],
            },
          ],
          "name": "ZodError",
        },
      ],
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_6 success (POST) 1`] = `
{
  "arr": [
    {
      "field2": {
        "enum1": "A",
        "nbr1": 123,
        "str1": "lala",
      },
      "query1": {
        "date": "2025-05-02T00:00:00.000Z",
        "nbr1": 999,
        "str1": "lalala",
      },
    },
    {
      "field2": {
        "enum1": "AA",
        "enum2": "Aahah",
      },
      "query1": {
        "date": "2025-10-02T00:00:00.000Z",
        "nbr1": 111,
        "str1": "lalala",
      },
    },
  ],
}
`;

exports[`AppController (e2e) /query_7 fail (POST) 1`] = `
{
  "issues": [
    {
      "code": "too_small",
      "exact": false,
      "inclusive": true,
      "message": "Number must be greater than or equal to -3",
      "minimum": -3,
      "path": [
        "nbr",
      ],
      "type": "number",
    },
    {
      "code": "invalid_string",
      "message": "Invalid email",
      "path": [
        "email",
      ],
      "validation": "email",
    },
    {
      "code": "invalid_string",
      "message": "Invalid url",
      "path": [
        "url",
      ],
      "validation": "url",
    },
    {
      "code": "invalid_string",
      "message": "Invalid",
      "path": [
        "phone",
      ],
      "validation": "regex",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_7 fail (POST) 2`] = `
{
  "issues": [
    {
      "code": "too_big",
      "exact": false,
      "inclusive": true,
      "maximum": 10,
      "message": "String must contain at most 10 character(s)",
      "path": [
        "str",
      ],
      "type": "string",
    },
    {
      "code": "too_big",
      "exact": false,
      "inclusive": true,
      "maximum": 23,
      "message": "Number must be less than or equal to 23",
      "path": [
        "nbr",
      ],
      "type": "number",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_7 success (POST) 1`] = `
{
  "email": "hey@hey.com",
  "nbr": 14,
  "phone": "1234567894",
  "str": "hello",
  "url": "https://google.com",
}
`;

exports[`AppController (e2e) /query_8 fail (POST) 1`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "nested",
        "long",
        "prop",
      ],
      "received": "nan",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_8 success (POST) 1`] = `
{
  "nested": {
    "long": {
      "prop": 21,
    },
  },
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 1`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "object",
      "message": "Required",
      "path": [
        "required",
      ],
      "received": "undefined",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 2`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "object",
      "message": "Required",
      "path": [
        "required",
        "long",
      ],
      "received": "undefined",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 3`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "object",
      "message": "Expected object, received number",
      "path": [
        "required",
        "long",
      ],
      "received": "number",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 4`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "required",
        "long",
        "prop",
      ],
      "received": "nan",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 5`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "object",
      "message": "Required",
      "path": [
        "opt",
        "long",
      ],
      "received": "undefined",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 6`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "opt",
        "long",
        "prop",
      ],
      "received": "nan",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 7`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "opt",
        "long",
        "prop",
      ],
      "received": "nan",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 fail (POST) 8`] = `
{
  "issues": [
    {
      "code": "invalid_type",
      "expected": "number",
      "message": "Expected number, received nan",
      "path": [
        "required",
        "opt",
        "prop",
      ],
      "received": "nan",
    },
  ],
  "name": "ZodError",
}
`;

exports[`AppController (e2e) /query_9 success (POST) 1`] = `
{
  "required": {
    "long": {
      "prop": 21,
    },
    "semi": {},
  },
}
`;

exports[`AppController (e2e) /query_9 success (POST) 2`] = `
{
  "required": {
    "long": {
      "prop": 21,
    },
  },
}
`;

exports[`AppController (e2e) /query_9 success (POST) 3`] = `
{
  "opt": {
    "long": {
      "prop": 21,
    },
  },
  "required": {
    "long": {
      "prop": 21,
    },
  },
}
`;
